### PR TITLE
codeintel: Trie-encode SCIP symbol names

### LINKED ISSUE #45609: ci/flake: GetBulkMonikerLocations disabled
### Test

GetBulkMonikerLocations

### Example failure

https://buildkite.com/sourcegraph/sourcegraph/builds/189044#_

### Disabling PR

https://github.com/sourcegraph/sourcegraph/pull/45608

### Additional details

_No response_

### PR DESCRIPTION
**Pull request by efritz**
Store symbol names in Postgres encoded as a forest of prefix tries rather than repeating the same very long prefixes that don't change within a given package/version pair.

On the write side: This PR reduces the total amount of data we have to send to Postgres during writes of SCIP indexes dramatically. This allows us to send more data per batch, and use less net bandwidth and reduce the number of round trips.

On the read side: The codeintel-qa suite (after the oobmigration runs) finishes all queries in ~20s and does not significantly alter latency from the previous behavior.

Also fixes #45609.

## Test plan

Updated unit tests, tested by hand.

### CODE REVIEW

File: internal/database/schema.codeintel.md [hunk 1]
```diff
@@ -166,20 +166,43 @@ Global metadatadata about a single processed upload.
...
+# Table "public.codeintel_scip_symbol_names"
+```
+    Column    |  Type   | Collation | Nullable | Default 
+--------------+---------+-----------+----------+---------
+ id           | integer |           | not null | 
+ upload_id    | integer |           | not null | 
+ name_segment | text    |           | not null | 
+ prefix_id    | integer |           |          | 
+Indexes:
+    "codeintel_scip_symbol_names_pkey" PRIMARY KEY, btree (upload_id, id)
```
Comment (efritz) [inline internal/database/schema.codeintel.md:178]: Need other indexes here

File: enterprise/internal/codeintel/uploads/internal/lsifstore/scip_write.go [hunk 2]
```diff
@@ -147,12 +149,33 @@ func (s *store) NewSymbolWriter(ctx context.Context, uploadID int) (SymbolWriter
...
+	symbolNames := make([]string, 0, len(symbolNameMap))
+	for symbolName := range symbolNameMap {
+		symbolNames = append(symbolNames, symbolName)
+	}
+	sort.Strings(symbolNames)
+
+	var symbolNameTrie trie.Trie
+	symbolNameTrie, s.nextID = trie.NewTrie(symbolNames, s.nextID)
+
+	// TODO - batch
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/internal/lsifstore/scip_write.go:172]: The batch/pre-calculate map things I will tackle in a follow-up PR as there are some other fallout effects I don't want to dirty this diff with.

File: enterprise/internal/oobmigration/migrations/codeintel/scip_migrator.go [hunk 3]
```diff
@@ -453,14 +456,17 @@ func makeSCIPWriter(ctx context.Context, tx *basestore.Store, uploadID int) (*sc
...
-	symbol_name text NOT NULL,
+	symbol_id integer NOT NULL,
 	document_lookup_id integer NOT NULL,
 	definition_ranges bytea,
 	reference_ranges bytea,
 	implementation_ranges bytea
 ) ON COMMIT DROP
 `
 
+// TODO :/
```
Comment (eseliger) [inline enterprise/internal/oobmigration/migrations/codeintel/scip_migrator.go:467]: :/ indeed

Comment (efritz) [reply enterprise/internal/oobmigration/migrations/codeintel/scip_migrator.go:467]: That's part of the batching comment in a follow-up PR.

Comment (efritz) [reply enterprise/internal/oobmigration/migrations/codeintel/scip_migrator.go:467]: For reference there's additional code in #45485 that will still hit main after this. I've just been cutting it up so it's easier for review.

**Pull request merged**